internal/fwschemadata/data_get_at_path_test.go

```diff
@@ -6536,6 +6536,249 @@ func TestDataGetAtPath(t *testing.T) {
+							Optional: true,
+							Type:     types.DynamicType,
+						},
+					},
+				},
+				TerraformValue: tftypes.NewValue(
+					tftypes.Object{
+						AttributeTypes: map[string]tftypes.Type{
+							"dynamic": tftypes.DynamicPseudoType,
+						},
+					},
+					map[string]tftypes.Value{
+						"dynamic": tftypes.NewValue(tftypes.DynamicPseudoType, nil),
+					},
+				),
+			},
+			path:     path.Root("dynamic"),
+			target:   new(types.Dynamic),
+			expected: pointer(types.DynamicNull()),
+		},
+		"DynamicType-types.dynamic-unknown": {
+			data: fwschemadata.Data{
+				Schema: testschema.Schema{
+					Attributes: map[string]fwschema.Attribute{
+						"dynamic": testschema.Attribute{
+							Optional: true,
+							Type:     types.DynamicType,
+						},
+					},
+				},
+				TerraformValue: tftypes.NewValue(
+					tftypes.Object{
+						AttributeTypes: map[string]tftypes.Type{
+							"dynamic": tftypes.DynamicPseudoType,
+						},
+					},
+					map[string]tftypes.Value{
+						"dynamic": tftypes.NewValue(tftypes.DynamicPseudoType, tftypes.UnknownValue),
+					},
+				),
+			},
+			path:     path.Root("dynamic"),
+			target:   new(types.Dynamic),
+			expected: pointer(types.DynamicUnknown()),
+		},
+		"DynamicType-types.dynamic-stringvalue": {
+			data: fwschemadata.Data{
+				Schema: testschema.Schema{
+					Attributes: map[string]fwschema.Attribute{
+						"string": testschema.Attribute{
+							Optional: true,
+							Type:     types.DynamicType,
+						},
+					},
+				},
+				TerraformValue: tftypes.NewValue(
+					tftypes.Object{
+						AttributeTypes: map[string]tftypes.Type{
+							"string": tftypes.String,
+						},
+					},
+					map[string]tftypes.Value{
+						"string": tftypes.NewValue(tftypes.String, "test"),
+					},
+				),
+			},
+			path:     path.Root("string"),
+			target:   new(types.Dynamic),
+			expected: pointer(types.DynamicValue(types.StringValue("test"))),
+		},
+		"DynamicType-types.dynamic-listvalue": {
+			data: fwschemadata.Data{
+				Schema: testschema.Schema{
+					Attributes: map[string]fwschema.Attribute{
+						"list": testschema.Attribute{
+							Optional: true,
+							Type:     types.DynamicType,
+						},
+					},
+				},
+				TerraformValue: tftypes.NewValue(
+					tftypes.Object{
+						AttributeTypes: map[string]tftypes.Type{
+							"list": tftypes.List{
+								ElementType: tftypes.String,
+							},
+						},
+					},
+					map[string]tftypes.Value{
+						"list": tftypes.NewValue(
+							tftypes.List{
+								ElementType: tftypes.String,
+							},
+							[]tftypes.Value{
+								tftypes.NewValue(tftypes.String, "test1"),
+								tftypes.NewValue(tftypes.String, "test2"),
+							},
+						),
+					},
+				),
+			},
+			path:   path.Root("list"),
+			target: new(types.Dynamic),
+			expected: pointer(types.DynamicValue(types.ListValueMust(
+				types.StringType,
+				[]attr.Value{
+					types.StringValue("test1"),
+					types.StringValue("test2"),
+				},
+			))),
```

Comment:
One interesting note that I wrote down...

If you have a known list value of type string in a dynamic type.... If you attempt to access an index that doesn't exist  or contains a known null value. It will not be `tftypes.String, nil`, it will be `tftypes.DynamicPseudoType, nil`. This is because `types.DynamicType` does not store any information about what kind of `attr.Type` might be stored in it.

This was only really a problem when I went to implement dynamic support for `ListType` and other collections (where it is much more important). I'm not sure if this will come up as a problem in other sections of framework code, but thought I would mention it

Reply:
> _Circling back to this comment_

I attempted to document this in all the collection type/value implementations for future maintainers (`basetypes.ListType`, `basetypes.ListValue`, etc). I did not run into any further issues and we added the proper validation to prevent this use-case from occurring. ⚡